lib/publiccloud/instance.pm

```diff
@@ -346,59 +355,88 @@ sub wait_for_ssh {
     # until it is not over we will recieve "ssh permission denied (pubkey)" error
     # but it is not good reason to die early because after it will be over DMS will return normal
     # user and error will be resolved.
-    $args{ignore_wrong_pubkey} //= 0;
+    #$args{ignore_wrong_pubkey} //= $args{proceed_on_failure};
     $args{username} //= $self->username();
     $args{public_ip} //= $self->public_ip();
+    $args{systemup_check} //= 1;    # check also system up, defult:true; when false,systemup check not executed,only ssh checked.
+    $args{logs} //= 1;    # log upload and info display
     my $start_time = time();
-    my $check_port = 1;
-    my $sleep_period = $args{ignore_wrong_pubkey} ? 20 : 1;
-
-    # Looping until reaching timeout or passing two conditions :
-    # - SSH port 22 is reachable
-    # - journalctl got message about reaching one of certain targets
-    while ((my $duration = time() - $start_time) < $args{timeout}) {
-        if ($check_port) {
-            $check_port = 0 if (script_run('nc -vz -w 1 ' . $self->{public_ip} . ' 22', quiet => 1) == 0);
-        }
-        else {
-            # On boottime test we do hard reboot which may change the instance address
-            script_run("ssh-keyscan $args{public_ip} | tee -a ~/.ssh/known_hosts") if (get_var('PUBLIC_CLOUD_CHECK_BOOT_TIME'));
-
-            my $output = $self->run_ssh_command(
-                cmd => 'sudo journalctl -b | grep -E "Reached target (Cloud-init|Default|Main User Target)"',
-                proceed_on_failure => 1,
-                username => $args{username});
-            if ($output =~ m/Reached target.*/) {
-                return $duration;
-            }
-            elsif ($output =~ m/Permission denied \(publickey\).*/) {
-                die "ssh permission denied (pubkey)" unless $args{ignore_wrong_pubkey};
-            }
-        }
-        sleep $sleep_period;
+    my $instance_msg = "instance: $self->{instance_id}, public IP: $self->{public_ip}";
+
+    # Looping until SSH port 22 is reachable or timeout.
+    # args{timeout} is the Total timeout,
+    # while the script_retry's timeout defines timeout of each retry iteration,
+    # however we use the nc command with option '-w 3' seconds of internal timeout.
+    # So, we calculate the retry number of the '-w' + delay consuming the Total timeout,
+    # but with 5sec timeout as upper limit of the command run:
+
+    my $exit_code = script_retry('nc -vz -w 3 ' . $self->{public_ip} . ' 22',
+        timeout => 5, delay => 1, retry => int($args{timeout} / 4) + 1, quiet => 1,
```

Comment:
The retry counter is off in my opinion. I think you wanted to do

```suggestion
        timeout => 5, delay => 1, retry => int($args{timeout} / 6) + 1, quiet => 1,
```

Why do I think this is wrong? See the following example:

    timeout = 120
    retry = (120/4)+1 = 31

But now, If we execute 31x a timeout of 5 seconds plus a delay of 1, then we get 186, which is not the expected timeout of 120.

If you want to quit after `timeout` seconds, then you need to calculate the retry counter by taking into account the full length of one retry interval, which is `timeout + delay`, not `timeout - delay`.

Reply:
code changed in last commit